src/webgpu/shader/execution/expression/expression.ts

```diff
@@ -485,23 +501,55 @@ function wgslInputVar(inputSource: InputSource, count: number) {
  */
 function wgslHeader(parameterTypes: Array<Type>, resultType: Type) {
   const usedF16 =
-    scalarTypeOf(resultType).kind === 'f16' ||
-    parameterTypes.some((ty: Type) => scalarTypeOf(ty).kind === 'f16');
-  const header = usedF16 ? 'enable f16;\n' : '';
+    scalarTypesOf(resultType).some(t => t.kind === 'f16') ||
+    parameterTypes.some((ty: Type) => scalarTypesOf(ty).some(t => t.kind === 'f16'));
+  let header = usedF16 ? 'enable f16;\n' : '';
+
+  const structTypes: StructType[] = [];
+  if (resultType instanceof StructType) {
+    structTypes.push(resultType);
+  }
+
+  structTypes.push(...(parameterTypes.filter(t => t instanceof StructType) as StructType[]));
+
+  // Passing through set to make sure added decls are unique
+  header += [...new Set(structTypes)].map(s => {
```

Comment:
This looks like it'll deduplicate structs if the JS objects are the same. Does this still work when serialising and deserializing a cached case, or do we potentially end up with multiple struct objects with the same name and members? I think it would be safer to use a map of struct name to struct.